### PR TITLE
Initial progress with Json Schema emitter

### PR DESCRIPTION
**Pull request by bterlson**
This PR implements an emitter for JSON Schema draft 2020-12. It supports all of JSON Schema with the exception of examples, dependentRequired, readOnly, and writeOnly. `examples` wants support for `valueOf` (#888).

This contains a breaking change: the default name for instantiated declarations is updated to capitalize template argument type names (so `BoxInt32` instead of `Boxint32`). I don't think this will cause any concern.

### CODE REVIEW

**Comment by bterlson** [comment]
Still some cleanups to do (e.g. considering not throwing errors but emitting diagnostics, validating inputs better in some cases), but I think this is ready for review.

File: packages/json-schema/lib/main.tsp
Comment (timotheeguerin) [inline packages/json-schema/lib/main.tsp:14]:
```suggestion
extern dec jsonSchema(target: unknown, baseUri?: string);
```
no?

File: packages/json-schema/samples/basic/ajv.js
Comment (timotheeguerin) [inline packages/json-schema/samples/basic/ajv.js:20]: Commented code

File: packages/json-schema/src/lib.ts
Comment (timotheeguerin) [inline packages/json-schema/src/lib.ts:40]:
```suggestion
  name: "@typespec/json-schema",
```

File: packages/json-schema/test/built-ins.test.ts
Comment (timotheeguerin) [inline packages/json-schema/test/built-ins.test.ts:27]: link to the offset issue?

File: packages/playground/vite.config.ts
Comment (timotheeguerin) [inline packages/playground/vite.config.ts:5]: Guess we should really have an emitter selector before we merge this

File: packages/json-schema/src/json-schema-emitter.ts
Comment (timotheeguerin) [inline packages/json-schema/src/json-schema-emitter.ts:618]: call `isAssignableTo`?

File: packages/json-schema/samples/basic/ajv.js
Comment (bterlson) [reply packages/json-schema/samples/basic/ajv.js:20]: Gonna delete this whole file, it's really just for testing purposes.

File: packages/json-schema/lib/main.tsp
Comment (bterlson) [reply packages/json-schema/lib/main.tsp:14]: Somehow the capital makes me happy here but... yes.

File: packages/playground/vite.config.ts
Comment (bterlson) [reply packages/playground/vite.config.ts:5]: Yeah I'm gonna revert all the playground changes before merge, the current thing is just a hack so you can play around with it in the meantime.

**Comment by bterlson** [comment]
Ok, I think this is ready for final review and merge. Some additional context:

1. Design meeting feedback was to handle duplicate identifiers. I've handled it by throwing an error.
2. Design meeting feedback was to consider creating declarations for Integers and questioning whether minValue/maxValue should be set depending on the integer type used. I have opted not to create declarations, if users want that behavior they can always declare it. I've also left minValue/maxValue in as I believe representing the allowed range of values for TypeSpec integer types is important. Happy to take feedback on this!
3. There were a few items of feedback that I want to defer to future development:
   1. Versioning decorators
   2. detecting non-sensical one-ofs
   3. @ extension support

File: common/changes/@typespec/json-schema/json-schema_2023-04-26-21-59.json
Comment (timotheeguerin) [inline common/changes/@typespec/json-schema/json-schema_2023-04-26-21-59.json:6]: Should this package also be pinned to the rest of the packages versions?

File: packages/compiler/emitter-framework/type-emitter.ts
Comment (timotheeguerin) [inline packages/compiler/emitter-framework/type-emitter.ts:674]: Should we be providing in the compiler or a library those naming helpers?

File: packages/json-schema/lib/main.tsp
Comment (timotheeguerin) [inline packages/json-schema/lib/main.tsp:21]:
```suggestion
extern dec baseUri(target: Reflection.Namespace, baseUri: valueof string);
```

File: packages/json-schema/lib/main.tsp
Comment (timotheeguerin) [inline packages/json-schema/lib/main.tsp:14]:
```suggestion
extern dec jsonSchema(target: unknown, baseUri?: valueof string);
```

File: packages/json-schema/lib/main.tsp
Comment (timotheeguerin) [inline packages/json-schema/lib/main.tsp:31]:
```suggestion
extern dec id(target: unknown, id: valueof string);
```

File: packages/json-schema/lib/main.tsp
Comment (timotheeguerin) [inline packages/json-schema/lib/main.tsp:38]:
```suggestion
extern dec multipleOf(target: numeric | Reflection.ModelProperty, value: valueof numeric);
```

File: packages/json-schema/lib/main.tsp
Comment (timotheeguerin) [inline packages/json-schema/lib/main.tsp:54]:
```suggestion
extern dec minContains(target: unknown[] | Reflection.ModelProperty, value: valueof int32);
```

File: packages/json-schema/lib/main.tsp
Comment (timotheeguerin) [inline packages/json-schema/lib/main.tsp:62]:
```suggestion
extern dec maxContains(target: unknown[] | Reflection.ModelProperty, value: valueof int32);
```

File: packages/json-schema/lib/main.tsp
Comment (timotheeguerin) [inline packages/json-schema/lib/main.tsp:74]:
```suggestion
extern dec minProperties(target: Record<unknown> | Reflection.ModelProperty, value: valueof int32);
```

File: packages/json-schema/lib/main.tsp
Comment (timotheeguerin) [inline packages/json-schema/lib/main.tsp:81]:
```suggestion
extern dec maxProperties(target: Record<unknown> | Reflection.ModelProperty, value: valueof int32);
```

File: packages/json-schema/lib/main.tsp
Comment (timotheeguerin) [inline packages/json-schema/lib/main.tsp:87]:
```suggestion
extern dec contentEncoding(target: string | Reflection.ModelProperty, value: valueof string);
```

File: packages/json-schema/lib/main.tsp
Comment (timotheeguerin) [inline packages/json-schema/lib/main.tsp:102]:
```suggestion
extern dec contentMediaType(target: string | Reflection.ModelProperty, value: valueof string);
```

**Review by timotheeguerin — COMMENTED** [review]
few things should be valueof

File: common/changes/@typespec/json-schema/json-schema_2023-04-26-21-59.json
Comment (bterlson) [reply common/changes/@typespec/json-schema/json-schema_2023-04-26-21-59.json:6]: Yup!

File: packages/compiler/emitter-framework/type-emitter.ts
Comment (bterlson) [reply packages/compiler/emitter-framework/type-emitter.ts:674]: Do you mean `Capitalize` here? Otherwise, this is a compiler/library since it's the emitter framework!

Comment (timotheeguerin) [reply packages/compiler/emitter-framework/type-emitter.ts:674]: yeah meant captialize but think we'd probably want pascalCase, camelCase, etc.. to
ah right yeah then maybe just export from the emitter framework, I can file an issue

File: rush.json
Comment (timotheeguerin) [inline rush.json:211]:
```suggestion
      "versionPolicyName": "typespec"
```

think need that if we want to pin it to the rest

File: packages/json-schema/src/lib.ts
Comment (timotheeguerin) [inline packages/json-schema/src/lib.ts:32]: I would copy the description here too so it shows up in the ref doc, which also make me think that we might want this to show up in the website no?

Comment (bterlson) [reply packages/json-schema/src/lib.ts:39]: It should show up on the website. As for manually written docs, I'll do a fast follow with those!

**Review by timotheeguerin — APPROVED** [review]

**Review by timotheeguerin — APPROVED** [review]

**Pull request merged**
